### PR TITLE
Add note regarding synchronous communication between too many isolates

### PR DESCRIPTION
**Pull request by aam**
Fixes https://github.com/dart-lang/sdk/issues/54687

### CODE REVIEW

File: src/content/language/isolates.md [hunk 1]
```diff
@@ -39,6 +39,12 @@ but here are some more situations where they can be useful:
 - Performing I/O, such as communicating with a database.
 - Handling a large volume of network requests.
 
+Also note a caveat due to a limit to the number of concurrently running isolates:
+
+ - The limit is not hardcoded to a particular number, it is calculated based on the Dart VM heap size available to the Dart application, can be considered to be between 8 and 32 depending on the platform.
+ - This limit doesn't affect asynchronous communction between isolates via messages - you can have hundreds of isolates running and making progress. The isolates are scheduled on the CPU in round-robin fashion and yield to each other often.
+ - Attempts to do *synchronous* communication between isolates over the limit though may result in a deadlock.
```
Comment (mkustermann) [inline src/content/language/isolates.md:46]: ... unless special care is taken (the C code that does synchronous communication would need to leave the current isolate before it blocks and re-enter it before returning to dart, see ...)

File: src/content/language/isolates.md [hunk 2]
```diff
@@ -39,6 +39,12 @@ but here are some more situations where they can be useful:
 - Performing I/O, such as communicating with a database.
 - Handling a large volume of network requests.
 
+Also note a caveat due to a limit to the number of concurrently running isolates:
```
Comment (mkustermann) [inline src/content/language/isolates.md:42]: I'd rephrase this as

"""
Synchronous blocking communication between isolates.    

If isolates want to communicate synchronously they have to leave the safety of Dart and resort to using C code via FFI to do so. Special care must be taken in this circumstance due to the Dart VM imposing a limit on the number of isolates that can run in parallel. If the number of isolates that synchronously block in FFI calls exceeds that limit no other isolate can run. If those blocking isolates wait on another unscheduled isolate to act, then deadlock can occur. To avoid this situation (preventing other isolates from running & possibly deadlock scenarios) such blocking C code must take care of leaving the current isolate before performing the blocking operation and re-entering the isolate before returning from the FFI call (see ...)
"""

(especially concurrently -> in parallel - we can run thousands of isolates concurrently but not in parallel, i.e. at the same time)

**Review by mkustermann — APPROVED** [review]

File: src/content/language/isolates.md (hunk 1 above)
Comment (aam) [reply src/content/language/isolates.md]: Thanks for the comment, I added this. PTAL.

File: src/content/language/isolates.md (hunk 2 above)
Comment (aam) [reply src/content/language/isolates.md]: This text sounds good, but did you mean for it to replace my version or you thought about adding it?

**Comment by aam** [comment]
not able to merge the PR, let's see if having Eric approve it allows merging. Or whether Eric can merge it.

**Review by MaryaBelanger — COMMENTED** [review]
I think it makes more sense to include this topic in the "[Limitations of isolates](https://dart.dev/language/concurrency#limitations-of-isolates)" section on the Concurrency > Overview page. 

The page titled "Isolates" that you have it on now is a walkthrough / example, whereas the Overview page is all the conceptual / background stuff. You can give your new content its own section under "Limitations of isolates", maybe titled "Synchronous blocking communication between isolates" like @mkustermann said in his comment. 

Sorry about the confusion! I can see now that it's not clear what the purpose of each page is, and that they don't really point to each other like they should. I'll fix that.

I haven't reviewed the actual content yet, but I'll come back if you agree with this change or want to discuss it more, thanks!

**Comment by aam** [comment]
> Edit: I can take over the PR to make this change too, if that's easier for you! Let me know :)

Didn't see the edit, so made the changes. Feel free to take over, thanks Marya!

File: src/content/language/concurrency.md [hunk 3]
```diff
@@ -398,6 +398,14 @@ isolate, it will remain untouched in the main isolate. This is how isolates are
...
 using isolates.
 
+#### Synchronous blocking communication between isolates
+
+There is a limit to the number of isolates running in parallel:
+
+ - The limit is not hardcoded to a particular number, it is calculated based on the Dart VM heap size available to the Dart application, can be considered to be between 8 and 32 depending on the platform.
+ - This limit doesn't affect asynchronous communction between isolates via messages - you can have hundreds of isolates running and making progress. The isolates are scheduled on the CPU in round-robin fashion and yield to each other often.
+ - Attempts to do *synchronous* communication between isolates over the limit though may result in a deadlock unless special care is taken (the C code that does synchronous communication would need to leave the current isolate before it blocks and re-enter it before returning to dart, see [`Dart_EnterIsolate`]({{site.repo.dart.sdk}}/blob/c9a8bbd8d6024e419b5e5f26b5131285eb19cc93/runtime/include/dart_api.h#L1254) and [`Dart_ExitIsolate`]({{site.repo.dart.sdk}}/blob/c9a8bbd8d6024e419b5e5f26b5131285eb19cc93/runtime/include/dart_api.h#L1455).
```
Comment (ericwindmill) [inline src/content/language/concurrency.md:407]: FWIW I would expand on this last bullet a bit. It's full of advanced concepts that aren't discussed in these docs at all, and they're concepts that many Dart developers probably have never thought about in their life. 

Given the recent user survey that found folks want more advanced isolate samples, we might even want to add example code of attempting synchronous communication between isolates.  That would probably constitute a new issue and separate PR. I'll leave that up to you, @MaryaBelanger, and let me know if you want to discuss.

**Comment by MaryaBelanger** [comment]
@parlough Can you checkout the links on `Dart_EnterIsolate` and `Dart_ExitIsolate` in the staged view? When I follow the links normally they work but in staging and serving the site locally they give 404/cannot be reached, respectively.

@aam Are there api docs pages for `Dart_EnterIsolate` and `Dart_ExitIsolate` on https://api.dart.dev/? I looked but couldn't find anything. Even if they're in dev and not stable, that works fine.

Also, I combined a little of Martin's comment with your work to provide more context in the entry. I don't think you can review your own PR, but let me know if you have any suggestions! Thanks

File: src/content/language/concurrency.md [hunk 4]
```diff
@@ -428,6 +428,30 @@ objects, so they're subject to the same limitations.
 [`Pointer`]: {{site.dart-api}}/{{site.sdkInfo.channel}}/dart-ffi/Pointer-class.html
 [`UserTag`]: {{site.dart-api}}/{{site.sdkInfo.channel}}/dart-developer/UserTag-class.html
 
+#### Synchronous blocking communication between isolates
+
+There is a limit to the number of isolates that can run in parallel, or *synchronously*.
```
Comment (aam) [inline src/content/language/concurrency.md:433]: I don't think `run... *synchronously*` makes sense. I would drop `, or *synchronously*`.

File: src/content/language/concurrency.md [hunk 5]
```diff
@@ -428,6 +428,30 @@ objects, so they're subject to the same limitations.
...
+This limit doesn't affect the standard *asynchronous* communication between isolates
+via messages in Dart. You can have hundreds of isolates running concurrently
+and making progress. The isolates are scheduled on the CPU in round-robin fashion,
+and yield to each other often.
+
+Isolates can only communicate *synchronously* outside of pure Dart,
+using C code via [FFI] to do so. 
+Attempts to perform synchronous communication between isolates in FFI calls
```
Comment (aam) [inline src/content/language/concurrency.md:441]: I would add: "...between isolates" by doing synchronous blocking "in FFI calls..".

File: src/content/language/concurrency.md [hunk 6]
```diff
@@ -428,6 +428,30 @@ objects, so they're subject to the same limitations.
...
+over the limit may result in deadlock unless special care is taken.
+The limit is not hardcoded to a particular number,
+it's calculated based on the Dart VM heap size available to the Dart application.
+
+To avoid this situation, the C code performing synchronous communication
```
Comment (aam) [inline src/content/language/concurrency.md:446]: synchronous blocking, rather synchronous communication

**Comment by aam** [comment]
> @aam Are there api docs pages for `Dart_EnterIsolate` and `Dart_ExitIsolate` on https://api.dart.dev/? I looked but couldn't find anything. Even if they're in dev and not stable, that works fine.

No, those API calls are not on api.dart.dev. They are part of  "Dart Embedding API Reference" with `dart_api.h` C++ header being documentation itself.

**Comment by MaryaBelanger** [comment]
Thanks for the feedback @aam. I'll merge once the builds done
